### PR TITLE
cli: Parse and allow hash value

### PR DESCRIPTION
**Pull request by fjahr**
This is an alternative approach to #15448 and also tries to eliminate "Error parsing JSON" errors. This came up in connection with `getblockstats` in this comment again: https://github.com/bitcoin/bitcoin/issues/19885#issuecomment-691192227 and that lead me to look into it.

Conceptually I am not completely against the approach of #15448 but it stalled and maybe this approach is favored by reviewers.

### CODE REVIEW

**Comment by fjahr** [comment]
Removed usage of `isxdigit` because the linter complained about locale dependency.

File: src/rpc/client.cpp
Comment (n-thumann) [inline src/rpc/client.cpp:225]: Maybe add that this method also accepts tx and blockhashes?
```suggestion
/** Non-RFC4627 JSON parser, accepts internal values (such as numbers, true, false, null or tx and blockhashes)
```

File: src/rpc/client.cpp
Comment (n-thumann) [inline src/rpc/client.cpp:231]:
```suggestion
    if ((!jVal.read(std::string("[")+strVal+std::string("]")) &&
        !jVal.read(std::string("[\"")+strVal+std::string("\"]"))) ||
```
Just as an idea, feel free to ignore :)
This would be shorter, requires no additional methods and also does the job (except for `getblockstats "'<blockhash>'"`). But - and that could be more important - it keeps the error message if the blockhash is e.g. too long (`hash_or_height must be of length 64 (not 66, for ...`, while your solution throws an JSON parsing error which could be misleading for a user 🤔

Comment (fjahr) [reply src/rpc/client.cpp:231]: If I see that correctly what this change will do is it will end up passing everything as a string to the server and never throw that error client-side. So basically the same as #15448 which removes `ParseNonRFCJSONValue` and only does validation server-side (the arguably better error message is coming from the server). The feedback to that was that the validation shouldn't be removed completely from the client-side which is the approach I am trying here.

File: src/rpc/client.cpp
Comment (fjahr) [reply src/rpc/client.cpp:225]: makes sense, I amended the comment.

Comment (n-thumann) [reply src/rpc/client.cpp:225]: Oh, I forgot: You probably want to change this in `src/rpc/client.h` as well for consistency :)

File: src/rpc/client.cpp
Comment (n-thumann) [reply src/rpc/client.cpp:231]: Sounds reasonable 👍

**Comment by n-thumann** [comment]
tACK https://github.com/bitcoin/bitcoin/commit/5076f8363eb658d24044dab4f618c414566f4b5f: ran functional & unit tests, works as expected. Only nit are the slightly degraded error messages when entering malformed hex strings ✌️

File: src/rpc/client.cpp
Comment (promag) [inline src/rpc/client.cpp:222]: Could use bool `IsHex(const std::string&)` from src/util/strencodings.h.

Comment (fjahr) [reply src/rpc/client.cpp:222]: Hm, my impression was that client.h/cpp should be stand-alone and not depend on other parts of the project? Maybe I am wrong about that but I thought I heard that somewhere and looking at the includes seemed to confirm it. Otherwise, I guess I could also just use `ParseHashV` from rpc/util.cpp?

Comment (promag) [reply src/rpc/client.cpp:222]: AFAIK it can't include server side components.

Comment (fjahr) [reply src/rpc/client.cpp]: Makes sense, changed to use `IsHex`.

File: src/rpc/client.cpp
Comment (fjahr) [reply src/rpc/client.cpp:222]: Done

**Comment by promag** [comment]
Does't feel right to place the workaround in `ParseNonRFCJSONValue`.

I prefer #15448.

**Comment by fjahr** [comment]
> Does't feel right to place the workaround in `ParseNonRFCJSONValue`.
> 
> I prefer #15448.

Not sure if that changes your mind but I pulled the check out of `ParseNonRFCJSONValue` which seems like a simpler change anyway.

@laanwj would you mind taking a quick look? This is mainly based on your feedback to #15448.

**Comment by laanwj** [comment]
I still don't like this kind of special casing based on "how a value looks". It adds a degree of ambiguity that can lead to nasty issues, especially if the value recognition becomes more complex. This is similar to what I said here: https://github.com/bitcoin/bitcoin/pull/15448#issuecomment-505495068

Still, I prefer this specialization to #15448, so if we really have to do something like this, mild ACK from me.

**Comment by ajtowns** [comment]
Could also change getblockstats to always expect a hash rather than hash_or_height, and use #16439 to convert "@height" into the hash on the client side.

**Comment by laanwj** [comment]
I like the idea of doing it on the client side. That alleviates my concerns which are mostly about programmatic use.

Although #20273 would be the most general solution, I guess this is shorter to type.

**Review by theStack — COMMENTED** [review]
Concept ACK

Are there currently any other RPCs than `getblockstats` that also benefit from this change? It may make sense to also update the RPC help examples of those to get rid of the quotation marks. E.g.:

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 04b9c6b1c..66e6b9af5 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -1746,7 +1746,7 @@ static UniValue getblockstats(const JSONRPCRequest& request) {RPCResult::Type::NUM, "utxo_size_inc", "The increase/decrease in size for the utxo index (not discounting op_return and similar)"},                                    }},
RPCExamples{
-                    HelpExampleCli("getblockstats", R"('"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"' '["minfeerate","avgfeerate"]')") +
+                    HelpExampleCli("getblockstats", R"(00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09 '["minfeerate","avgfeerate"]')") +
                     HelpExampleCli("getblockstats", R"(1000 '["minfeerate","avgfeerate"]')") +                                                                                                  
```

**Comment by fjahr** [comment]
> Concept ACK
> 
> Are there currently any other RPCs than `getblockstats` that also benefit from this change? It may make sense to also update the RPC help examples of those to get rid of the quotation marks. E.g.:
> 
> ```diff
> diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
> index 04b9c6b1c..66e6b9af5 100644
> --- a/src/rpc/blockchain.cpp
> +++ b/src/rpc/blockchain.cpp
> @@ -1746,7 +1746,7 @@ static UniValue getblockstats(const JSONRPCRequest& request) {RPCResult::Type::NUM, "utxo_size_inc", "The increase/decrease in size for the utxo index (not discounting op_return and similar)"},                                    }},
> RPCExamples{
> -                    HelpExampleCli("getblockstats", R"('"00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09"' '["minfeerate","avgfeerate"]')") +
> +                    HelpExampleCli("getblockstats", R"(00000000c937983704a73af28acdec37b049d214adbda81d7e2a3dd146f6ed09 '["minfeerate","avgfeerate"]')") +
>                      HelpExampleCli("getblockstats", R"(1000 '["minfeerate","avgfeerate"]')") +                                                                                                  
> ```

Yepp, it's is also useful now in `gettxoutsetinfo` after some changes were merged there in the meantime. I don't know of any others. I added a commit with these changes and rebased for good measures.

**Review by aureleoules — APPROVED** [review]
ACK bf12f7c3de49d226d1d282948f91fb8e47e37bb9
I think this makes the interface less confusing to use.

**Review by stickies-v — COMMENTED** [review]
Approach NACK

It's an elegant approach, and I think we should strive for more intuitive interfaces and this PR definitely would achieve that for the cli. However I think adding these kinds of exception handling 1) increases the maintenance burden instead of fixing the problem at its source - good method/endpoint design and 2) encourage further suboptimally designed methods/endpoints because we now have the tooling to interface with them more conveniently, making it more difficult to fix in the future.

For example, in this case I'd prefer the approach of letting the methods deal with parameter polymorphism (if they have a strong enough reason to use it in the first place), and encourage developers to be conservative with complex parameter design. I've updated `getblockheight` (but similar approach should probably work for gettxoutsetinfo` even though I didn't look at it) in the below diff to showcase how that could work. It's a very quick patch so please let me know if I didn't consider some negative side effects of this approach.

<details>
<summary>git diff on top of bf12f7c3d</summary>

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index f3cab7483..6bd86a5d8 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -109,17 +109,17 @@ static const CBlockIndex* ParseHashOrHeight(const UniValue& param, ChainstateMan
     LOCK(::cs_main);
     CChain& active_chain = chainman.ActiveChain();
 
-    if (param.isNum()) {
-        const int height{param.getInt<int>()};
-        if (height < 0) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d is negative", height));
+    const auto height{ToIntegral<int>(param.get_str())};
+    if (height) {
+        if (*height < 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d is negative", *height));
         }
         const int current_tip{active_chain.Height()};
-        if (height > current_tip) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d after current tip %d", height, current_tip));
+        if (*height > current_tip) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Target block height %d after current tip %d", *height, current_tip));
         }
 
-        return active_chain[height];
+        return active_chain[*height];
     } else {
         const uint256 hash{ParseHashV(param, "hash_or_height")};
         const CBlockIndex* pindex = chainman.m_blockman.LookupBlockIndex(hash);
@@ -1715,7 +1715,7 @@ static RPCHelpMan getblockstats()
                 "\nCompute per block statistics for a given window. All amounts are in satoshis.\n"
                 "It won't work for some heights with pruning.\n",
                 {
-                    {"hash_or_height", RPCArg::Type::NUM, RPCArg::Optional::NO, "The block hash or height of the target block", "", {"", "string or numeric"}},
+                    {"hash_or_height", RPCArg::Type::STR, RPCArg::Optional::NO, "The block hash or height of the target block", "", {"", "string or numeric"}},
                     {"stats", RPCArg::Type::ARR, RPCArg::DefaultHint{"all values"}, "Values to plot (see result below)",
                         {
                             {"height", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "Selected statistic"},
diff --git a/src/rpc/client.cpp b/src/rpc/client.cpp
index 90c69db7d..be47c3b79 100644
--- a/src/rpc/client.cpp
+++ b/src/rpc/client.cpp
@@ -158,7 +158,6 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "listdescriptors", 0, "private" },
     { "verifychain", 0, "checklevel" },
     { "verifychain", 1, "nblocks" },
-    { "getblockstats", 0, "hash_or_height" },
     { "getblockstats", 1, "stats" },
     { "pruneblockchain", 0, "height" },
     { "keypoolrefill", 0, "newsize" },
```
</details>

**Comment by fjahr** [comment]
I am not interested enough in this anymore to argue about implementation details. Could be marked as up-for-grabs if the issue remains unsolved.

**Pull request closed without merging**
